File: src/intensity/utility.cpp

/*
 *=============================================================================
 * Copyright (C) 2008 Alon Zakai ('Kripken') kripkensteiner@gmail.com
 *
 * This file is part of the Intensity Engine project,
 *    http://www.intensityengine.com
 *
 * The Intensity Engine is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * The Intensity Engine is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with the Intensity Engine.  If not, see
 *     http://www.gnu.org/licenses/
 *     http://www.gnu.org/licenses/agpl-3.0.html
 *=============================================================================
 */


#include <string>
#include <map>

#include "cube.h"

#include "utility.h"

using namespace boost;


//==================================
// Utilities. Many of these will
// simply use boost/python for now,
// but we can replace them with
// suitable portable specific
// things later
//==================================


std::string Utility::SHA256(std::string text)
{
    EXEC_PYTHON("import hashlib");

    REFLECT_PYTHON_ALTNAME(hashlib.sha256, hashlib_sha256);

    return python::extract<std::string>( hashlib_sha256(text).attr("hexdigest")() );
}

//==============================
// String utils
//==============================

std::string Utility::toString(std::string val)
{
    return val;
}

#define TO_STRING(type)                  \
std::string Utility::toString(type val)  \
{                                        \
    std::stringstream ss;                \
    std::string ret;                     \
    ss << val;                           \
    return ss.str();                     \
}

TO_STRING(int)
TO_STRING(long)
TO_STRING(double)

bool Utility::validateAlphaNumeric(std::string input, std::string alsoAllow)
{
    EXEC_PYTHON("import re");

    REFLECT_PYTHON_ALTNAME( re.escape, re_escape);

    python::object original = python::object(input);
    for (unsigned int i = 0; i < alsoAllow.size(); i++)
    {
        original = original.attr("replace")(alsoAllow[i], ""); // Ignore the alsoAllow chars
    }

    python::object test = re_escape(original);

    bool ret = (test == original); // After escaping all non-alphanumeric, must be no change

    if (!ret)
        Logging::log(Logging::WARNING, "Validation of %s failed (using alphanumeric + %s)\r\n", input.c_str(), alsoAllow.c_str());

    return ret;
}

bool Utility::validateNotContaining(std::string input, std::string disallow)
{
    python::object original = python::object(input);
    int index = python::extract<int>(original.attr("find")(disallow));
    return index == -1; // -1 means it wasn't found, which is ok
}

bool Utility::validateRelativePath(std::string input)
{
    REFLECT_PYTHON( validate_relative_path );
    return python::extract<bool>(validate_relative_path(input));
}

std::string Utility::readFile(std::string name)
{
    REFLECT_PYTHON( open );
    try
    {
        python::object data = open(name, "r").attr("read")();
        return python::extract<std::string>(data);
    }
    catch(boost::python::error_already_set const &)
    {
        printf("Error in Python execution of readFile\r\n");
        PyErr_Print();
        assert(0 && "Halting on Python error");
    }

}


//==============================
// Config file utilities
//==============================

// Caches for speed; no need to call Python every time
typedef std::map<std::string, std::string> ConfigCacheString;
typedef std::map<std::string, int> ConfigCacheInt;
typedef std::map<std::string, float> ConfigCacheFloat;

ConfigCacheString configCacheString;
ConfigCacheInt configCacheInt;
ConfigCacheFloat configCacheFloat;

inline std::string getCacheKey(std::string section, std::string option)
    { return section + "|" + option; };

#define GET_CONFIG(type, Name, python_type)                                           \
type pythonGet##Name(std::string section, std::string option, type defaultVal)        \
{                                                                                     \
    Logging::log(Logging::DEBUG, "Config cache fail, going to Python: %s/%s\r\n",     \
                 section.c_str(), option.c_str());                                    \
    REFLECT_PYTHON(get_config);                                                       \
    REFLECT_PYTHON_ALTNAME(python_type, ptype);                                       \
    return python::extract<type>( ptype( get_config(section, option, defaultVal) ) ); \
} \
\
type Utility::Config::get##Name(std::string section, std::string option, type defaultVal)\
{ \
    std::string cacheKey = getCacheKey(section, option);\
    ConfigCache##Name::iterator iter = configCache##Name.find(cacheKey);\
    if (iter != configCache##Name.end())\
        return iter->second;\
    else {\
        type value = pythonGet##Name(section, option, defaultVal); \
        configCache##Name[cacheKey] = value; \
        return value; \
    } \
}

GET_CONFIG(std::string, String, str)
GET_CONFIG(int,         Int,    int)
GET_CONFIG(float,       Float,  float)

#define SET_CONFIG(type, Name)                                                  \
void Utility::Config::set##Name(std::string section, std::string option, type value) \
{                                                                               \
    REFLECT_PYTHON(set_config);                                                 \
    set_config(section, option, value);                                         \
    std::string cacheKey = getCacheKey(section, option); \
    configCache##Name[cacheKey] = value; \
}

SET_CONFIG(std::string, String)
SET_CONFIG(int,         Int)
SET_CONFIG(float,       Float)


// Cubescript accessibility

void get_config(char *section, char *option)
{
    result(Utility::Config::getString(section, option, "?").c_str());
}

COMMAND(get_config, "ss");


//==============================
// System Info
//==============================

extern int clockrealbase;

int Utility::SystemInfo::currTime()
{
#ifdef SERVER
    return enet_time_get();
#else // CLIENT
    return SDL_GetTicks() - clockrealbase;
#endif
// This old method only changes during calls to updateworld etc.!
//    extern int lastmillis;
//    return lastmillis; // We wrap around the sauer clock
}

File: src/python/intensity/utility.py
"""
Various utilities.
"""


#=============================================================================
# Copyright (C) 2008 Alon Zakai ('Kripken') kripkensteiner@gmail.com
#
# This file is part of the Intensity Engine project,
#    http://www.intensityengine.com
#
# The Intensity Engine is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, version 3.
#
# The Intensity Engine is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with the Intensity Engine.  If not, see
#     http://www.gnu.org/licenses/
#     http://www.gnu.org/licenses/agpl-3.0.html
#=============================================================================

import os
from math import atan2, pi, sqrt, asin
import hashlib

from intensity.base import *
from intensity.logging import *


## Lets 'print' be used from boost::python as a function. For debugging. Will not be needed in Python 3.0.
def print_func(param):
    print param

## A simple timer that counts down time.
class Timer:
    def __init__(this, seconds):
        this.seconds = seconds

    def tick(this, seconds):
        this.seconds -= seconds
        return this.seconds <= 0


## A timer that resets itthis: after the allotted time to be counted down is done, it starts counting again.
class ResettingTimer:
    """doctest unit testing:

    >>> rt = ResettingTimer(10)
    >>> import time

    >>> rt.tick(4)
    False

    You should not do the following, but it shows what is happening internally:
    >>> rt.seconds
    6

    >>> rt.tick(4)
    False

    >>> rt.tick(4)
    True

    Again, this should not be done, but shows what happened internally:
    >>> rt.seconds
    10

    As you can see, the timer reset itthis.
    """
    ## @param seconds The amount of seconds to count down each time.
    def __init__(this, seconds):
        this.max_seconds = seconds
        this.seconds     = seconds

    ## Increment the time by a certain amount.
    ## @param seconds The amount of seconds to increment the timer by.
    ## @return Whether the counting down has finished. If so, the timer also resets itthis to a new count down.
    def tick(this, seconds):
        this.seconds -= seconds
        if this.seconds <= 0:
            this.seconds = this.max_seconds
            return True
        else:
            return False


## Removes old backup files from a directory, based on their modification times
## @param path The directory to do the cleaning in
## @param suffix The suffix that defines a backup file (other files are ignored) - without the '.' character
## @param to_leave The number of backup files to leave - all others are removed
def clean_up_backups(path, suffix, to_leave):
    names = filter(lambda name: name[-(len(suffix)+1):] == '.' + suffix, os.listdir(path))
    infos = [ (name, os.stat( os.path.join(path, name) ).st_mtime) for name in names ]
    infos.sort(key = lambda pair: pair[1])
    for i in range(len(infos)-to_leave):
#        print os.path.join(path, infos[i][0])
        os.remove( os.path.join(path, infos[i][0]) )


if Global.CLIENT:
    ## CEGUI interface
    def run_lua_function(name):
        CAPI.run_lua_function(name)


def validate_relative_path(path):
    '''
    >>> validate_relative_path('packages/models/cannon/barrel/../skin.jpg')
    True
    >>> validate_relative_path('skin.jpg')
    False
    >>> validate_relative_path('../skin.jpg')
    False
    >>> validate_relative_path('packages/../skin.jpg')
    False
    >>> validate_relative_path('packages/skin.jpg')
    True
    >>> validate_relative_path('packages//../skin.jpg')
    False
    >>> validate_relative_path('packages//skin.jpg')
    True
    >>> validate_relative_path('packages/models/../../skin.jpg')
    False
    >>> validate_relative_path('packages/models/../skin.jpg')
    True
    '''
    path = path.replace('\\', '/') # Use entirely UNIX-style seps to check
    level = -1 # So first addition puts us in 0, a valid level
    for seg in path.split('/'):
        if seg == '..':
            level -= 1
        elif seg != '.' and seg != '':
            level += 1
        if level < 0:
            return False
    level -= 1 # Last segment, the file itself, doesn't count
    return level >= 0


__shown_process_script_warning = False
def process_script(script):
    '''
    Optimize a JavaScript script before we run it.
    We comment out lines that look like log(XXX, ...);
    if XXX would not be shown anyhow. That is, if
    we have
        log(DEBUG, serializeJSON(something));
    then the expensive function call will be done
    even if the output is not shown. So by commenting
    it out, we prevent that.
    '''
    if get_config('Logging', 'optimize_scripts', '1') != '1': return script

    global __shown_process_script_warning
    if not __shown_process_script_warning:
        log(logging.WARNING, 'Optimizing scripts by commenting out unneeded loggings')
        __shown_process_script_warning = True

    ret = []
    for line in script.split('\n'):
        line = line.strip()
        for i in range(len(logging.strings)):
            prefix = 'log(' + logging.strings[i]
            if line[0:len(prefix)] == prefix:
                if line[-2:] == ');':
                    if not logging.should_show(i):
                        line = ''
                else:
                    log(logging.WARNING, 'script optimization not sure about:' + line)
                break
        ret.append(line)

    return '\n'.join(ret)

def check_newer_than(principal, *others):
    '''
    Checks if the file 'principal' is newer than
    various other files. If the principal doesn't
    exist, that is also consider not being newer than.
    '''

    if not os.path.exists(principal):
        return False

    mtime = os.stat(principal).st_mtime
    for other in others:
        if os.stat(other).st_mtime > mtime: # == means nothing to worry about
            return False

    return True

def prepare_texture_replace(filename, curr_textures):
    '''
    Given a file of old textures actually used in a map (from listtex), and a list of current texture names, makes a lookup table.
    The lookup table, given an index in the old textures, returns the new correct index for it using the current textures.

    Method:
        1. Run the old map with the old map.js script
        2. Do /listtex and save the output printed to the console to a file.
            (You may need to remove some unneeded textures at the end.)
        2. Fix the map.js to set up textures the way you want.
        3. Run with that new map.js.
        4. Do /massreplacetex FILENAME with the name of the file you created before
    '''
#    print "CURR:", curr_textures
    old_data = open(filename, 'r').read().replace('\r', '').split('\n')
#    print "OLD:", old_data
    lookup = {}
    for item in old_data:
        if item == '': continue
        index, name = item.split(' : ')
        index = int(index)
#        print "    ", index, name,
        # find new index - the index of that same name in the current textures
        lookup[index] = curr_textures.index(name)
        print lookup[index]
        curr_textures[lookup[index]] = 'zzz' # Do not find this again. This makes it ok to have the same texture twice,
                                                # with different rotations - we will process them ok

#    print "LOOKUP:", lookup
    return lookup


if __name__ == '__main__':
    import doctest
    doctest.testmod()

